### PR TITLE
[flang] Add basic -mtune support

### PR DESCRIPTION
**Pull request by AlexisPerry**
This PR adds -mtune as a valid flang flag and passes the information through to LLVM IR as an attribute on all functions.  No specific architecture optimizations are added at this time.

### CODE REVIEW

File: mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td
Comment (Dinistro) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:68]: Dead code?

File: clang/lib/Driver/ToolChains/Flang.cpp
Comment (Dinistro) [inline clang/lib/Driver/ToolChains/Flang.cpp:417]: Nit: `strcmp` is a C function. I assume that `A->getValue()` returns a StringRef, or something else that supports normal `==` comparison.

**Review by Dinistro — COMMENTED** [review]
This is missing tests for the LLVM import and export. I suspect that this is currently still part of the function's passthrough dictionary, which should be changed.

File: flang/tools/bbc/bbc.cpp
Comment (tarunprabhu) [inline flang/tools/bbc/bbc.cpp:374]: What needs to be fixed here? If it needs to be deferred to a later patch, a more detailed inline comment would be helpful.

**Review by tarunprabhu — COMMENTED** [review]
Is there a way to test that the attribute is lowered to FIR and perhaps also on to LLVM? `flang/test/Lower/` has some tests which check for the presence of certain constructs in the LLVM IR. Could something be written similar to those?

File: flang/tools/bbc/bbc.cpp
Comment (AlexisPerry) [reply flang/tools/bbc/bbc.cpp:374]: Thank you for catching that.  I meant to remove that comment as the thing it was referring to has been fixed.  I'll clean it up now.

File: clang/lib/Driver/ToolChains/Flang.cpp
Comment (AlexisPerry) [reply clang/lib/Driver/ToolChains/Flang.cpp:417]: I see your point, however when I build the code using normal `==` comparison I get a warning:
`warning: comparison with string literal results in unspecified behavior`
and I am inclined to avoid such possible unspecified behavior.

**Review by banach-space — COMMENTED** [review]
Thanks for working on this @AlexisPerry ! Could you add some tests?

File: clang/lib/Driver/ToolChains/Flang.cpp
Comment (tarunprabhu) [reply clang/lib/Driver/ToolChains/Flang.cpp:417]: `A` is an `llvm::Arg`. The `getValue()` method [returns](https://llvm.org/doxygen/classllvm_1_1opt_1_1Arg.html#ac268590692356db84db78050196b4940) a `const char*`, so unless it is wrapped in a `llvm::StringRef`, the `==` operator cannot be used for string comparison.

It is surprising though. I, too, would have expected it to return a `llvm::StringRef`.

File: flang/test/Lower/tune-cpu-fir.f90
Comment (tarunprabhu) [inline flang/test/Lower/tune-cpu-fir.f90:1]: Why do you have the first `target-cpu` check on its own? Is it intended to check for the _absence_ of the `tune-cpu` attribute when `target-cpu` is provided on the command line? If so, there should be a `CPU-NOT` annotation as well. Otherwise, it is odd to test for `target-cpu` when this file is intended to test `tune-cpu`. 

The third test which checks the case when both `target -cpu` and `tune-cpu` are provided is good.

File: flang/test/Lower/tune-cpu-llvm.f90
Comment (tarunprabhu) [inline flang/test/Lower/tune-cpu-llvm.f90:1]: Consider replacing `#0` with `#{{[0-9]+}}`. While the attributes are almost certainly going to be in `#0` for such a small function, you don't actually _require_ this to match `#0`, so you it might be better to let it match any number. 

Also, if the `tune-cpu` attribute is expected to be on the same line, you could use `CHECK-SAME`.

File: flang/test/Lower/tune-cpu-fir.f90
Comment (AlexisPerry) [reply flang/test/Lower/tune-cpu-fir.f90:1]: Thank you for the catch! Yes, the CPU-NOT line was missing.  I have rectified this.

File: flang/test/Lower/tune-cpu-llvm.f90
Comment (AlexisPerry) [reply flang/test/Lower/tune-cpu-llvm.f90:1]: Those are great suggestions, thank you.  I have made the appropriate updates.

File: flang/tools/bbc/bbc.cpp
Comment (tarunprabhu) [inline flang/tools/bbc/bbc.cpp:374]: NIT: Would it be worth creating a `constexpr const char* tuneCPU = ""` and using that in place of the empty string literal here? It might make it clear what that parameter is. 

I don't have a strong opinion on this one way or another. If you or anyone else has a different opinion, I'm happy to defer to it.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1802]: Could you also add a small test that verifies the import / export works independent from Flang?

The target_cpu attributes are tested here:
```
llvm-project/mlir/test/Target/LLVMIR/target-cpu.mlir
llvm-project/mlir/test/Target/LLVMIR/Import/target-cpu.ll
```
It may make sense to add the tests in the same files and maybe rename them to target-and-tune-cpu.mlir/ll or have a separate set of files if the attributes are unrelated (which is not my impression).

Thanks!

File: flang/include/flang/Frontend/TargetOptions.h
Comment (AlexisPerry) [reply flang/include/flang/Frontend/TargetOptions.h]: Thank you for the suggestion.  I have made the change. :-)

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (AlexisPerry) [reply mlir/lib/Target/LLVMIR/ModuleImport.cpp:1802]: Thank you so much for this suggestion and for the pointers for how to do this kind of test.  I was struggling to figure out the proper mechanism and you helped greatly.  I have added two separate tests for tune-cpu instead of combining with the target-cpu ones because I think it's cleaner, but I am happy to combine if you prefer. Just let me know.  Thanks again for your help!

File: flang/tools/bbc/bbc.cpp
Comment (AlexisPerry) [reply flang/tools/bbc/bbc.cpp]: Thank you Tarun, I think that's a much better way to do this. I have made the change.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleImport.cpp:1802]: Nice!

Separate tests are fine.

**Review by tarunprabhu — COMMENTED** [review]
Thanks for the changes. LGTM, but wait for @banach-space to approve.

File: clang/lib/Driver/ToolChains/Flang.cpp
Comment (Dinistro) [reply clang/lib/Driver/ToolChains/Flang.cpp:417]: A few lines further up, the string literal get's wrapped into a `StringRef`, which seems to work. 

`if (A->getValue() == StringRef{"Accelerate"})`

File: mlir/test/Target/LLVMIR/Import/tune-cpu.ll
Comment (Dinistro) [inline mlir/test/Target/LLVMIR/Import/tune-cpu.ll:4]: Can you add a check that ensures that this is not also placed in the passthrough dictionary? I don't see any part in the code that would avoid this. Adding it to the `ExplicitAttributes` set in `ModuleImport.cpp` should be enough.

File: flang/test/Lower/tune-cpu-fir.f90
Comment (banach-space) [inline flang/test/Lower/tune-cpu-fir.f90:1]: This is probably a border-line case, but I would consider this a "driver" rather than a "lowering" test. I'm biased though 😅

**Review by banach-space — APPROVED** [review]
LGTM, thanks for implementing this 🙏🏻

File: mlir/test/Target/LLVMIR/Import/tune-cpu.ll
Comment (AlexisPerry) [reply mlir/test/Target/LLVMIR/Import/tune-cpu.ll:4]: From what I can see, L1685 of ModuleImport.cpp has tune-cpu listed among the ExplicitAttributes.  Is this sufficient?  Or is further checking needed?  Thank you so much.

File: clang/lib/Driver/ToolChains/Flang.cpp
Comment (AlexisPerry) [reply clang/lib/Driver/ToolChains/Flang.cpp:417]: Thank you for that!  I have made the requested update and there was no build warning.

File: flang/test/Lower/tune-cpu-fir.f90
Comment (AlexisPerry) [reply flang/test/Lower/tune-cpu-fir.f90:1]: I hear you.  I modeled this test closely after target-features-*.f90 in that same directory, which is why I put it in flang/test/Lower.  I am willing to move it to flang/test/Driver, but if I do so should I also move the target-features-*.f90 tests?  Thank you so much.

**Comment by AlexisPerry** [comment]
I have lost my upstream commit privileges due to inactivity, so once this is fully approved, could someone merge it on my behalf?  Thank you.

File: mlir/test/Target/LLVMIR/Import/tune-cpu.ll
Comment (Dinistro) [reply mlir/test/Target/LLVMIR/Import/tune-cpu.ll:4]: Given that we are both not entirely sure what happens, it might be sensible to check for that, just to be sure 🙂

**Review by Dinistro — APPROVED** [review]
The MLIR side looks good to me % one remaining nit comment. Thanks for addressing the comments.

File: flang/test/Lower/tune-cpu-fir.f90
Comment (banach-space) [reply flang/test/Lower/tune-cpu-fir.f90:1]: > if I do so should I also move the target-features-*.f90 tests?

Yes, but to me that would qualify as an "unrelated change" (i.e. sth for a separate PR, no need to worry about it here). 

In general, this PR is about enabling a flag in Flang and making sure that it actually works. This kind of "integration" is a driver task, hence my suggestion.

File: mlir/test/Target/LLVMIR/Import/tune-cpu.ll
Comment (AlexisPerry) [reply mlir/test/Target/LLVMIR/Import/tune-cpu.ll:4]: Alright, after some digging, I can confirm that `tune-cpu` is listed among the `kExplicitAttributes` and that these attributes are expressly excluded from being added to the passthrough dictionary based on the code in `processPassthroughAttrs` in `ModuleImport.cpp`.

File: flang/test/Lower/tune-cpu-fir.f90
Comment (AlexisPerry) [reply flang/test/Lower/tune-cpu-fir.f90:1]: Thanks for explaining more thoroughly.  I have moved the test to the Driver directory as requested.

**Comment by AlexisPerry** [comment]
I believe I've addressed all the review comments and all the checks have passed.  Could someone with commit access please merge this on my behalf?  Thank you.

**Comment by banach-space** [comment]
> LLVM Buildbot has detected a new failure on builder `flang-aarch64-libcxx` running on `linaro-flang-aarch64-libcxx` while building `clang,flang,mlir` at step 6 "test-build-unified-tree-check-flang".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/89/builds/791
> 
> Here is the relevant piece of the build log for the reference:
> 
> ```
> Step 6 (test-build-unified-tree-check-flang) failure: test (failure)
> ******************** TEST 'Flang :: Lower/tune-cpu-llvm.f90' FAILED ********************
> Exit Code: 2
> 
> Command Output (stderr):
> --
> RUN: at line 1: /home/tcwg-buildbot/worker/flang-aarch64-libcxx/build/bin/flang-new -mtune=pentium4 -S -emit-llvm /home/tcwg-buildbot/worker/flang-aarch64-libcxx/llvm-project/flang/test/Lower/tune-cpu-llvm.f90 -o - | /home/tcwg-buildbot/worker/flang-aarch64-libcxx/build/bin/FileCheck /home/tcwg-buildbot/worker/flang-aarch64-libcxx/llvm-project/flang/test/Lower/tune-cpu-llvm.f90
> + /home/tcwg-buildbot/worker/flang-aarch64-libcxx/build/bin/flang-new -mtune=pentium4 -S -emit-llvm /home/tcwg-buildbot/worker/flang-aarch64-libcxx/llvm-project/flang/test/Lower/tune-cpu-llvm.f90 -o -
> + /home/tcwg-buildbot/worker/flang-aarch64-libcxx/build/bin/FileCheck /home/tcwg-buildbot/worker/flang-aarch64-libcxx/llvm-project/flang/test/Lower/tune-cpu-llvm.f90
> flang-new: error: unsupported argument 'pentium4' to option '-mtune='
> FileCheck error: '<stdin>' is empty.
> FileCheck command line:  /home/tcwg-buildbot/worker/flang-aarch64-libcxx/build/bin/FileCheck /home/tcwg-buildbot/worker/flang-aarch64-libcxx/llvm-project/flang/test/Lower/tune-cpu-llvm.f90
> 
> --
> 
> ********************
> ```

Most likely https://github.com/llvm/llvm-project/pull/95043/files#diff-a29a79ef4763ed66987d979a7a8a4ff87d242101fe133d5188577b9ff144b805 requires X86 to be enabled. Could somebody either fix or revert? I don’t have access to Git ATM.

**Pull request merged**
